erpnext/www/book-appointment/index.py

```diff
@@ -0,0 +1,166 @@
+
+
+@frappe.whitelist(allow_guest=True)
+def get_appointment_settings():
+    settings = frappe.get_doc('Appointment Booking Settings')
+    return settings
+
+
+@frappe.whitelist(allow_guest=True)
+def is_enabled():
+    enable_scheduling = frappe.db.get_single_value(
+        'Appointment Booking Settings', 'enable_scheduling')
+    return enable_scheduling
+
+
+@frappe.whitelist(allow_guest=True)
+def get_holiday_list(holiday_list_name):
+    holiday_list = frappe.get_doc('Holiday List', holiday_list_name)
+    return holiday_list
+
+
+@frappe.whitelist(allow_guest=True)
+def get_timezones():
+    return pytz.all_timezones
+
+
+@frappe.whitelist(allow_guest=True)
+def get_appointment_slots(date, timezone):
+    import pytz
+    guest_timezone = pytz.timezone(timezone)
+    format_string = '%Y-%m-%d %H:%M:%S'
+    query_start_time = datetime.datetime.strptime(
+        date + ' 00:00:00', format_string)
+    query_end_time = datetime.datetime.strptime(
+        date + ' 23:59:59', format_string)
+    local_timezone = frappe.utils.get_time_zone()
+    local_timezone = pytz.timezone(local_timezone)
+    query_start_time = guest_timezone.localize(query_start_time)
+    query_end_time = guest_timezone.localize(query_end_time)
+    query_start_time = query_start_time.astimezone(local_timezone)
+    query_end_time = query_end_time.astimezone(local_timezone)
+    now = datetime.datetime.now()
+    # now = local_timezone.localize(now)
+    # Database queries
+    settings = frappe.get_doc('Appointment Booking Settings')
+    holiday_list = frappe.get_doc('Holiday List', settings.holiday_list)
+    timeslots = get_available_slots_between(
+        query_start_time, query_end_time, settings)
+
+    # Filter timeslots based on date
+    converted_timeslots = []
+    for timeslot in timeslots:
+        timeslot = local_timezone.localize(timeslot)
+        timeslot = timeslot.astimezone(guest_timezone)
+        timeslot = timeslot.replace(tzinfo=None)
+        # Check if holiday
+        if _is_holiday(timeslot.date(), holiday_list):
+            converted_timeslots.append(
+                dict(time=timeslot, availability=False))
+            continue
+        # Check availability
+        if check_availabilty(timeslot, settings) and timeslot >= now:
+            converted_timeslots.append(
+                dict(time=timeslot, availability=True))
+        else:
+            converted_timeslots.append(
+                dict(time=timeslot, availability=False))
+    date_required = datetime.datetime.strptime(
+        date + ' 00:00:00', format_string).date()
+    converted_timeslots = filter_timeslots(date_required, converted_timeslots)
+    return converted_timeslots
+
+
+def get_available_slots_between(query_start_time, query_end_time, settings):
+    records = _get_records(query_start_time, query_end_time, settings)
+    timeslots = []
+    appointment_duration = datetime.timedelta(
+        minutes=settings.appointment_duration)
```

Comment:
Is this tested for minutes greater than 60?

Reply:
Yes, it's handled by the python datetime library.